wemo_base.py

```diff
@@ -0,0 +1,111 @@
+        self._thread = None
+        self._discovering = False
+        self._updating = False
+
+    def configure(self, context):
+        super().configure(context)
+        self._thread = spawn(self._discover)
+
+    def process_signals(self, signals):
+        if not self.device:
+            self.logger.warning(
+                'No WeMo device connected, dropping {} signals'.format(
+                    len(signals)))
+            if self._discovering:
+                return
+            else:
+                self._thread = spawn(self._discover)
+                return
+        outgoing_signals = []
+        for signal in signals:
+            new_signal = self.get_output_signal(
+                self.execute_wemo_command(signal), signal)
+            outgoing_signals.append(new_signal)
+        self.notify_signals(outgoing_signals)
+
+    def execute_wemo_command(self, signal):
+        """ Override this method - do something to the wemo device
+
+        Return a result to be enriched on to the incoming signal
+        """
+        return {}
+
+    def rediscover(self):
+        self.logger.info('Rediscover command received!')
+        if self._discovering:
+            status = 'Discovery already in progress'
+        else:
+            status = 'OK'
+            if self.device:
+                status += ', dropped device \"{}\" with MAC {}'\
+                    .format(self.device.name, self.device.mac)
+            self.device = None
+            self._thread = spawn(self._discover)
+        self.logger.info(status)
+        return {'status': status}
+
+    def is_valid_device(self, device):
+        """ Override to determine whether the device should be considered
+
+        Can check device type, device MAC address, etc
+
+        This parent function will check that the MAC address matches if it
+        was specified in the block parameters
+
+        Return a boolean for whether it is valid
+        """
+        if self.device_mac():
+            return device.mac == self.device_mac()
+        # If no MAC specified consider it valid
+        return True
+
+    def _discover(self):
+        self._discovering = True
+        self.device = None
+        while not self.device:
+            self.logger.debug('Discovering WeMo devices on network...')
+            try:
+                devices = discover_devices()
+            except ConnectionError:
+                self.logger.error('Error discovering devices, aborting')
+                self._discovering = False
+                return
+            self.logger.debug('Found {} WeMo devices'.format(len(devices)))
+            for device in devices:
+                self.logger.debug('Checking device {}'.format(device))
+                if self.is_valid_device(device):
+                    self.device = device
+                    break
+            else:
+                # If we didn't find a device wait a bit before trying again
+                sleep(0.1)
```

Comment:
Is this where you saw the block loop at maximum effort when you lost network? `discover_devices` is a blocking call, but i suppose if there's no network it may return immediately. In that case, shouldn't this sleep actually be longer?

Reply:
I'm looking through and actually not sure why I added it haha, it must have just felt right to me at the time. I wrote most of this WemoSwitch code and did the "base" roll out a few months back, I finally got around to tidying and committing the other day though. So my memory of it isn't as fresh as to why I did it, but I think it makes sense here.

I bumped it up to one second too, that feels more reasonable to me now that I think about it more.